File: Contract.Compiler.Expressions/ConstantFolding.fs
namespace Contract.Compiler.Expressions.Internal

open Contract.Compiler.AST
open Contract.Compiler.Parsing

/// <summary>
/// Compile-time constant folding (FEATURE_PROPOSALS §15).
/// Folds literals, arithmetic, comparison, logical operators, and references
/// to constants declared earlier in the same contract.
/// </summary>
[<AutoOpen>]
module ConstantFolding =

    let private constValueTypeName (value: obj option) =
        match value with
        | None -> null
        | Some (:? int) -> "int"
        | Some (:? double) -> "double"
        | Some (:? string) -> "string"
        | Some (:? bool) -> "bool"
        | _ -> null

    let private constantTypeMatches (annotation: string) (value: obj option) =
        let actual = constValueTypeName value
        if actual = null then true // null folds / unknown kinds: skip
        else annotation = actual

    /// <summary>Folds a binary operator applied to two constant values.</summary>
    let rec foldConstantBinary (op: string) (left: obj option) (right: obj option) : obj option * bool =
        match op with
        | "==" | "!=" ->
            let equal =
                match left, right with
                | Some (:? int as a), Some (:? double as b) -> (double a) = b
                | Some (:? double as a), Some (:? int as b) -> a = (double b)
                | Some (:? int as a), Some (:? int as b) -> a = b
                | Some (:? double as a), Some (:? double as b) -> a = b
                | Some (:? string as a), Some (:? string as b) -> a = b
                | Some (:? bool as a), Some (:? bool as b) -> a = b
                | None, None -> true
                | _ -> false
            let result = if op = "==" then equal else not equal
            (Some (box result), true)

        | "<" | "<=" | ">" | ">=" ->
            let cmp =
                match left, right with
                | Some (:? int as a), Some (:? int as b) -> a.CompareTo(b)
                | Some (:? double as a), Some (:? double as b) -> a.CompareTo(b)
                | Some (:? int as a), Some (:? double as b) -> (double a).CompareTo(b)
                | Some (:? double as a), Some (:? int as b) -> a.CompareTo(double b)
                | Some (:? string as a), Some (:? string as b) -> a.CompareTo(b)
                | _ -> 0
            let result =
                match op with
                | "<" -> cmp < 0
                | "<=" -> cmp <= 0
                | ">" -> cmp > 0
                | ">=" -> cmp >= 0
                | _ -> false
            (Some (box result), true)

        | "+" ->
            match left, right with
            | Some (:? int as a), Some (:? int as b) -> (Some (box (a + b)), true)
            | Some (:? double as a), Some (:? double as b) -> (Some (box (a + b)), true)
            | Some (:? int as a), Some (:? double as b) -> (Some (box (double a + b)), true)
            | Some (:? double as a), Some (:? int as b) -> (Some (box (a + double b)), true)
            | Some (:? string as a), Some (:? string as b) -> (Some (box (a + b)), true)
            | _ -> (None, false)

        | "-" ->
            match left, right with
            | Some (:? int as a), Some (:? int as b) -> (Some (box (a - b)), true)
            | Some (:? double as a), Some (:? double as b) -> (Some (box (a - b)), true)
            | Some (:? int as a), Some (:? double as b) -> (Some (box (double a - b)), true)
            | Some (:? double as a), Some (:? int as b) -> (Some (box (a - double b)), true)
            | _ -> (None, false)

        | "*" ->
            match left, right with
            | Some (:? int as a), Some (:? int as b) -> (Some (box (a * b)), true)
            | Some (:? double as a), Some (:? double as b) -> (Some (box (a * b)), true)
            | Some (:? int as a), Some (:? double as b) -> (Some (box (double a * b)), true)
            | Some (:? double as a), Some (:? int as b) -> (Some (box (a * double b)), true)
            | _ -> (None, false)

        | "/" ->
            match left, right with
            | Some (:? int as a), Some (:? int as b) when b <> 0 -> (Some (box (a / b)), true)
            | Some (:? double as a), Some (:? double as b) when b <> 0.0 -> (Some (box (a / b)), true)
            | Some (:? int as a), Some (:? double as b) when b <> 0.0 -> (Some (box (double a / b)), true)
            | Some (:? double as a), Some (:? int as b) when b <> 0 -> (Some (box (a / double b)), true)
            | _ -> (None, false)

        | "%" ->
            match left, right with
            | Some (:? int as a), Some (:? int as b) when b <> 0 -> (Some (box (a % b)), true)
            | Some (:? double as a), Some (:? double as b) when b <> 0.0 -> (Some (box (a % b)), true)
            | Some (:? int as a), Some (:? double as b) when b <> 0.0 -> (Some (box (double a % b)), true)
            | Some (:? double as a), Some (:? int as b) when b <> 0 -> (Some (box (a % double b)), true)
            | _ -> (None, false)

        | _ -> (None, false)

    /// <summary>
    /// Folds a compile-time constant expression. Returns true when the
    /// expression is a foldable constant and sets <c>value</c>.
    /// </summary>
    let rec tryFoldConstant
        (ctx: ParserContext)
        (expr: Expression)
        (contract: ContractDeclaration)
        (host: IParserHost)
        : bool * obj option =
        match expr with
        | :? LiteralExpression as lit -> (true, Some lit.Value)

        | :? IdentifierExpression as id ->
            let field =
                contract.Fields
                |> Seq.tryFind (fun f -> f.IsConst && f.IsStatic && f.Name = id.Name)
            match field with
            | None -> (false, None)
            | Some f -> (true, f.ConstantValue |> Option.ofObj)

        | :? UnaryExpression as unary ->
            let (ok, operand) = tryFoldConstant ctx unary.Operand contract host
            if not ok then (false, None)
            else
                match unary.Operator, operand with
                | "-", Some (:? int as i) -> (true, Some (box (-i)))
                | "-", Some (:? double as d) -> (true, Some (box (-d)))
                | "!", Some (:? bool as b) -> (true, Some (box (not b)))
                | _ -> (false, None)

        | :? BinaryExpression as bin ->
            // Short-circuit forms
            if bin.Operator = "&&" || bin.Operator = "||" then
                let (okL, lhs) = tryFoldConstant ctx bin.Left contract host
                if not okL then (false, None)
                else
                    match lhs with
                    | Some (:? bool as leftBool) ->
                        if (bin.Operator = "&&" && not leftBool) || (bin.Operator = "||" && leftBool) then
                            let result = bin.Operator = "&&" |> not
                            (true, Some (box result))
                        else
                            let (okR, rhs) = tryFoldConstant ctx bin.Right contract host
                            if not okR then (false, None)
                            else
                                match rhs with
                                | Some (:? bool as rightBool) ->
                                    let result =
                                        if bin.Operator = "&&" then leftBool && rightBool
                                        else leftBool || rightBool
                                    (true, Some (box result))
                                | _ -> (false, None)
                    | _ -> (false, None)
            else
                let (okL, left) = tryFoldConstant ctx bin.Left contract host
                let (okR, right) = tryFoldConstant ctx bin.Right contract host
                if not okL || not okR then (false, None)
                else
                    let (value, ok) = foldConstantBinary bin.Operator left right
                    (ok, value)

        | _ -> (false, None)

    /// <summary>
    /// Best-effort type inference from an expression for contract-scope
    /// lambda desugaring. Returns null when the type cannot be inferred.
    /// </summary>
    let rec inferTypeFromExpression (expr: Expression) : string =
        match expr with
        | :? LiteralExpression as lit ->
            match lit.Value with
            | :? int -> "int"
            | :? double -> "double"
            | :? string -> "string"
            | :? bool -> "bool"
            | _ -> null
        | :? BinaryExpression as b -> inferTypeFromExpression b.Left
        | :? UnaryExpression as u -> inferTypeFromExpression u.Operand
        | :? IdentifierExpression as id -> id.Name
        | _ -> null

File: Contract.Compiler/Parsing/Expressions/ConstantFolding.cs
using System;
using System.Linq;
using Contract.Compiler.AST;
using Contract.Compiler.Parsing;

namespace Contract.Compiler.Expressions.Internal
{
    /// <summary>
    /// Compile-time constant folding (FEATURE_PROPOSALS section 15).
    /// Folds literals, arithmetic, comparison, logical operators, and references
    /// to constants declared earlier in the same contract.
    /// </summary>
    internal static class ConstantFolding
    {
        private static string? ConstValueTypeName(object? value)
        {
            return value switch
            {
                int => "int",
                double => "double",
                string => "string",
                bool => "bool",
                _ => null
            };
        }

        private static bool ConstantTypeMatches(string annotation, object? value)
        {
            var actual = ConstValueTypeName(value);
            if (actual == null) return true; // null folds / unknown kinds: skip
            return annotation == actual;
        }

        /// <summary>Folds a binary operator applied to two constant values.</summary>
        public static (bool ok, object? value) FoldConstantBinary(string op, object? left, object? right)
        {
            switch (op)
            {
                case "==":
                case "!=":
                {
                    bool equal;
                    switch (left, right)
                    {
                        case (int a, double b): equal = (double)a == b; break;
                        case (double a, int b): equal = a == (double)b; break;
                        case (int a, int b): equal = a == b; break;
                        case (double a, double b): equal = a == b; break;
                        case (string a, string b): equal = a == b; break;
                        case (bool a, bool b): equal = a == b; break;
                        case (null, null): equal = true; break;
                        default: equal = false; break;
                    }
                    return (op == "==" ? equal : !equal, true);
                }

                case "<":
                case "<=":
                case ">":
                case ">=":
                {
                    int cmp;
                    switch (left, right)
                    {
                        case (int a, int b): cmp = a.CompareTo(b); break;
                        case (double a, double b): cmp = a.CompareTo(b); break;
                        case (int a, double b): cmp = ((double)a).CompareTo(b); break;
                        case (double a, int b): cmp = a.CompareTo((double)b); break;
                        case (string a, string b): cmp = a.CompareTo(b); break;
                        default: cmp = 0; break;
                    }
                    bool result = op switch
                    {
                        "<" => cmp < 0,
                        "<=" => cmp <= 0,
                        ">" => cmp > 0,
                        ">=" => cmp >= 0,
                        _ => false
                    };
                    return (true, (object?)result);
                }

                case "+":
                    switch (left, right)
                    {
                        case (int a, int b): return (true, (object?)(a + b));
                        case (double a, double b): return (true, (object?)(a + b));
                        case (int a, double b): return (true, (object?)((double)a + b));
                        case (double a, int b): return (true, (object?)(a + (double)b));
                        case (string a, string b): return (true, (object?)(a + b));
                        default: return (false, null);
                    }

                case "-":
                    switch (left, right)
                    {
                        case (int a, int b): return (true, (object?)(a - b));
                        case (double a, double b): return (true, (object?)(a - b));
                        case (int a, double b): return (true, (object?)((double)a - b));
                        case (double a, int b): return (true, (object?)(a - (double)b));
                        default: return (false, null);
                    }

                case "*":
                    switch (left, right)
                    {
                        case (int a, int b): return (true, (object?)(a * b));
                        case (double a, double b): return (true, (object?)(a * b));
                        case (int a, double b): return (true, (object?)((double)a * b));
                        case (double a, int b): return (true, (object?)(a * (double)b));
                        default: return (false, null);
                    }

                case "/":
                    switch (left, right)
                    {
                        case (int a, int b) when b != 0: return (true, (object?)(a / b));
                        case (double a, double b) when b != 0.0: return (true, (object?)(a / b));
                        case (int a, double b) when b != 0.0: return (true, (object?)((double)a / b));
                        case (double a, int b) when b != 0: return (true, (object?)(a / (double)b));
                        default: return (false, null);
                    }

                case "%":
                    switch (left, right)
                    {
                        case (int a, int b) when b != 0: return (true, (object?)(a % b));
                        case (double a, double b) when b != 0.0: return (true, (object?)(a % b));
                        case (int a, double b) when b != 0.0: return (true, (object?)((double)a % b));
                        case (double a, int b) when b != 0: return (true, (object?)(a % (double)b));
                        default: return (false, null);
                    }

                default:
                    return (false, null);
            }
        }

        /// <summary>
        /// Folds a compile-time constant expression. Returns true when the
        /// expression is a foldable constant and sets <c>value</c>.
        /// </summary>
        public static (bool ok, object? value) TryFoldConstant(
            ParserContext ctx, Expression expr,
            ContractDeclaration contract, IParserHost host)
        {
            switch (expr)
            {
                case LiteralExpression lit:
                    return (true, lit.Value);

                case IdentifierExpression id:
                {
                    var field = contract.Fields.FirstOrDefault(f => f.IsConst && f.IsStatic && f.Name == id.Name);
                    return field != null ? (true, (object?)field.ConstantValue) : (false, null);
                }

                case UnaryExpression unary:
                {
                    var (ok, operand) = TryFoldConstant(ctx, unary.Operand, contract, host);
                    if (!ok) return (false, null);
                    return (unary.Operator, operand) switch
                    {
                        ("-", int i) => (true, (object?)(-i)),
                        ("-", double d) => (true, (object?)(-d)),
                        ("!", bool b) => (true, (object?)(!b)),
                        _ => (false, null)
                    };
                }

                case BinaryExpression bin:
                {
                    // Short-circuit forms
                    if (bin.Operator == "&&" || bin.Operator == "||")
                    {
                        var (okL, lhs) = TryFoldConstant(ctx, bin.Left, contract, host);
                        if (!okL) return (false, null);
                        if (lhs is bool leftBool)
                        {
                            if ((bin.Operator == "&&" && !leftBool) || (bin.Operator == "||" && leftBool))
                            {
                                bool result = bin.Operator != "&&";
                                return (true, (object?)result);
                            }
                            var (okR, rhs) = TryFoldConstant(ctx, bin.Right, contract, host);
                            if (!okR) return (false, null);
                            if (rhs is bool rightBool)
                            {
                                bool result = bin.Operator == "&&" ? leftBool && rightBool : leftBool || rightBool;
                                return (true, (object?)result);
                            }
                            return (false, null);
                        }
                        return (false, null);
                    }

                    var (okLeft, left) = TryFoldConstant(ctx, bin.Left, contract, host);
                    var (okRight, right) = TryFoldConstant(ctx, bin.Right, contract, host);
                    if (!okLeft || !okRight) return (false, null);
                    return FoldConstantBinary(bin.Operator, left, right);
                }

                default:
                    return (false, null);
            }
        }

        /// <summary>
        /// Best-effort type inference from an expression for contract-scope
        /// lambda desugaring. Returns null when the type cannot be inferred.
        /// </summary>
        public static string? InferTypeFromExpression(Expression expr)
        {
            return expr switch
            {
                LiteralExpression lit => lit.Value switch
                {
                    int => "int",
                    double => "double",
                    string => "string",
                    bool => "bool",
                    _ => null
                },
                BinaryExpression b => InferTypeFromExpression(b.Left),
                UnaryExpression u => InferTypeFromExpression(u.Operand),
                IdentifierExpression id => id.Name,
                _ => null
            };
        }
    }
}
